File: com/bubblesort/gemini/gemini.java
package com.bubblesort.gemini;

import java.util.Random;

public class gemini {
    public static void sort(int[] arr) {
        int n = arr.length;
        boolean swapped;
    
        // Loop through the array n-1 times
        for (int i = 0; i < n - 1; i++) {
          swapped = false;
    
          // Compare adjacent elements and swap if they are in the wrong order
          for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
              swap(arr, j, j + 1);
              swapped = true;
            }
          }
    
          // If we haven't made any swaps in this iteration, the array is already sorted
          if (!swapped) {
            break;
          }
        }
      }
    
      // Helper method to swap two elements in the array
      private static void swap(int[] arr, int i, int j) {
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
      }
  public static void main(String[] args) {
    // Criando um array com 1000 valores
    int[] array = new int[1000];

    // Preenchendo o array com valores aleatórios entre 1 e 100
    Random random = new Random();
    for (int i = 0; i < array.length; i++) {
        array[i] = random.nextInt(100) + 1; // Gera um número aleatório entre 1 e 100
    }

    // Chamando o método de ordenação
    sort(array);
  }
}

File: Python/bubblesort/gemini/gemini.py
def bubble_sort(arr):
  """
  Implements the Bubble Sort algorithm with optimization.

  Args:
    arr: The list to be sorted.

  Returns:
    The sorted list.
  """

  n = len(arr)

  # Traverse through all array elements
  for i in range(n):
    # Last i elements are already in place
    swapped = False
    for j in range(0, n - i - 1):

      # Traverse the array from 0 to n-i-1
      # Swap if the element found is greater
      # than the next element
      if arr[j] > arr[j + 1]:
        arr[j], arr[j + 1] = arr[j + 1], arr[j]
        swapped = True

    # If no two elements were swapped by inner loop, then break
    if not swapped:
      break

  return arr

import random

def main():
    num_iterations = 1000
    array_size = 1000  # Tamanho do array a ser ordenado

    for _ in range(num_iterations):
        # Gera um array aleatório com mil valores
        arr = [random.randint(1, 1000) for _ in range(array_size)]
        # Ordena o array
        bubble_sort(arr)

if __name__ == "__main__":
    main()